src/main/java/com/shapesecurity/shift/es2016/codegen/location/CodeGenWithLocation.java

```diff
@@ -0,0 +1,70 @@
+package com.shapesecurity.shift.es2016.codegen.location;
+
+import com.shapesecurity.functional.F2;
+import com.shapesecurity.functional.data.Maybe;
+import com.shapesecurity.shift.es2016.ast.Module;
+import com.shapesecurity.shift.es2016.ast.Node;
+import com.shapesecurity.shift.es2016.ast.Script;
+import com.shapesecurity.shift.es2016.codegen.CodeRep;
+import com.shapesecurity.shift.es2016.parser.SourceLocation;
+import com.shapesecurity.shift.es2016.parser.SourceSpan;
+import com.shapesecurity.shift.es2016.reducer.Director;
+import com.shapesecurity.shift.es2016.reducer.Reducer;
+import com.shapesecurity.shift.es2016.reducer.WrappedReducer;
+import org.jetbrains.annotations.NotNull;
+
+public class CodeGenWithLocation extends WrappedReducer<CodeRep> {
+	@NotNull
+	protected LocationMeta meta;
+
+	private CodeGenWithLocation(@NotNull CtorArgs args) {
+		super(args.wrap, args.reducer);
+		this.meta = args.meta;
+	}
+
+	private static class CtorArgs {
+		// this is just to deal with stupidity in the compiler
```

Comment:
You probably want more rationale than this. Also, I think we may be able to replace this with a static helper method.